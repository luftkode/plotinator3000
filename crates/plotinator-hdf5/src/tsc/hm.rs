use anyhow::{bail, ensure};
use ndarray::{ArrayBase, Dim, OwnedRepr};
use plotinator_log_if::prelude::RawPlot;

// 6-dimensional array: [N, 413, 2, 76, 6, 4]
type HmArray = ArrayBase<OwnedRepr<i64>, Dim<[usize; 6]>>;

/// Wrapper around the hm dataset from `/RX/monomial_basis_data/hm`
pub(crate) struct HmData<'h5> {
    h5file: &'h5 hdf5::File, // Keep file open for future access
    dataset_name: String,    // Path to dataset
    shape: [usize; 6],       // Only load shape initially
    inner: Option<HmArray>,  // Load full data lazily when needed
}

impl<'h5> HmData<'h5> {
    pub fn from_hdf5(h5: &'h5 hdf5::File) -> hdf5::Result<Self> {
        let dataset_name = "/RX/monomial_basis_data/hm".to_owned();
        let dataset = h5.dataset(&dataset_name)?;
        let shape_vec = dataset.shape();
        let shape: [usize; 6] = shape_vec
            .try_into()
            .map_err(|e| hdf5::Error::Internal(format!("Dataset is not 6D: {e:?}")))?;

        Ok(Self {
            h5file: h5,
            dataset_name,
            shape,
            inner: None, // don't read data yet
        })
    }

    pub fn shape(&self) -> &[usize] {
        self.shape.as_ref()
    }

    /// Load full data only if needed
    #[allow(
        dead_code,
        reason = "We will need this later, and want to test that the functionality doesn't break"
    )]
    pub fn load_full(&mut self) -> hdf5::Result<()> {
        if self.inner.is_none() {
            let dataset = self.h5file.dataset(&self.dataset_name)?;
            let hm_data = dataset.read::<i64, _>()?;
            self.inner = Some(hm_data);
        }
        Ok(())
    }

    #[allow(
        dead_code,
        reason = "We will need this later, and want to test that the functionality doesn't break"
    )]
    /// Create time-aligned point series from GPS timestamps
    /// Returns Vec<[f64; 2]> where each element is [timestamp, `hm_value`]
    /// The first dimension of hm data should match (or be 1 less than) GPS marks count
    pub fn create_time_series(
        &self,
        gps_timestamps: &[f64],
        coords: [usize; 5], // [dim1, dim2, dim3, dim4, dim5] - first dim varies with time
    ) -> anyhow::Result<Vec<[f64; 2]>> {
        let hm_len = self.shape()[0]; // First dimension length
        let gps_len = gps_timestamps.len();

        // Validate alignment: hm length should match GPS or be 1 less
        ensure!(
            hm_len == gps_len || hm_len == gps_len - 1,
            format!(
                "HM data length ({hm_len}) doesn't align with GPS timestamps ({gps_len}). Expected {gps_len} or {}",
                gps_len - 1
            )
        );

        let mut time_series = Vec::with_capacity(hm_len);
        let [dim1, dim2, dim3, dim4, dim5] = coords;

        for (i, gps_ts) in gps_timestamps.iter().enumerate().take(hm_len) {
            if let Some(hm_value) = self.get([i, dim1, dim2, dim3, dim4, dim5]) {
                time_series.push([*gps_ts, *hm_value as f64]);
            } else {
                bail!(
                    "Invalid HM coordinates at index {i}: [{i}, {dim1}, {dim2}, {dim3}, {dim4}, {dim5}]"
                );
            }
        }

        Ok(time_series)
    }

    #[allow(
        dead_code,
        reason = "We will need this later, and want to test that the functionality doesn't break"
    )]
    /// Access element at specific coordinates [dim0, dim1, dim2, dim3, dim4, dim5]
    /// Shape is [N, 413, 2, 76, 6, 4]
    pub fn get(&self, coords: [usize; 6]) -> Option<&i64> {
        self.inner.as_ref()?.get(coords)
    }

    #[allow(
        dead_code,
        reason = "We will need this later, and want to test that the functionality doesn't break"
    )]
    /// Get a slice along the first dimension at fixed other coordinates
    /// Example: `get_slice_dim0([0, 1, 1, 1, 1])`
    pub fn get_slice_dim0(&self, coords: [usize; 5]) -> Option<ndarray::ArrayView1<i64>> {
        let [dim1, dim2, dim3, dim4, dim5] = coords;
        self.inner
            .as_ref()?
            .slice(ndarray::s![.., dim1, dim2, dim3, dim4, dim5])
            .into_dimensionality()
            .ok()
    }

    // Build metadata without loading full data
    pub fn build_plots_and_metadata(
        &self,
        gps_timestamps: &[f64],
    ) -> (Vec<RawPlot>, Vec<(String, String)>) {
        let hm_len = self.shape()[0];
        let gps_len = gps_timestamps.len();
        let metadata = vec![
            ("HM length".to_owned(), hm_len.to_string()),
            ("GPS length".to_owned(), gps_len.to_string()),
        ];
        (vec![], metadata)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use plotinator_test_util::test_file_defs::tsc::*;
    use testresult::TestResult;

    #[test]
    fn read_hm_data() -> TestResult {
        let h5file = hdf5::File::open(tsc())?;
        let mut hm_data = HmData::from_hdf5(&h5file)?;

        println!("HM Data shape: {:?}", hm_data.shape());

        // Test accessing specific elements (triggers lazy load)
        hm_data.load_full()?; // optional, can rely on get() lazy load
        if let Some(value) = hm_data.get([0, 0, 0, 0, 0, 0]) {
            println!("Element [0,0,0,0,0,0]: {value}");
        }

        // Test accessing a slice
        if let Some(slice) = hm_data.get_slice_dim0([0, 0, 0, 0, 0]) {
            println!("Slice along first dimension: {slice:?}");
        }

        // Take a small sample for snapshot testing to avoid huge output
        let sample_coords = [
            [0, 0, 0, 0, 0, 0],
            [0, 1, 0, 0, 0, 0],
            [1, 0, 1, 0, 0, 0],
            [2, 10, 0, 10, 0, 0],
            [3, 100, 1, 50, 5, 3],
        ];

        let mut sample_values = Vec::new();
        for coords in sample_coords {
            if let Some(value) = hm_data.get(coords) {
                sample_values.push((coords, *value));
            }
        }

        insta::assert_debug_snapshot!(sample_values);
        Ok(())
    }

    #[test]
    fn test_hm_data_dimensions() -> TestResult {
        let h5file = hdf5::File::open(tsc())?;
        let mut hm_data = HmData::from_hdf5(&h5file)?;
        hm_data.load_full()?;

        let gps_timestamps = vec![0.0, 1.0, 2.0, 3.0];

        let (_plots, metadata) = hm_data.build_plots_and_metadata(&gps_timestamps);
        let shape = hm_data.shape();

        assert_eq!(metadata.len(), 2);

        // Verify expected shape [4, 413, 2, 76, 6, 4]
        assert_eq!(shape.len(), 6, "Should be 6-dimensional");
        assert_eq!(shape[0], 4, "First dimension should be 4");
        assert_eq!(shape[1], 413, "Second dimension should be 413");
        assert_eq!(shape[2], 2, "Third dimension should be 2");
        assert_eq!(shape[3], 76, "Fourth dimension should be 76");
        assert_eq!(shape[4], 6, "Fifth dimension should be 6");
        assert_eq!(shape[5], 4, "Sixth dimension should be 4");

        // Verify we can access boundary elements
        assert!(hm_data.get([3, 412, 1, 75, 5, 3]).is_some());
        assert!(hm_data.get([4, 0, 0, 0, 0, 0]).is_none()); // Out of bounds

        Ok(())
    }
}
