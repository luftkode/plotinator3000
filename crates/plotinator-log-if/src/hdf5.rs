use std::path::Path;

use crate::prelude::Plotable;

/// A given HDF5 file should implement this
pub trait SkytemHdf5: Plotable + Sized {
    /// Take a path to an HDF5 file and parse it
    fn from_path(path: impl AsRef<Path>) -> anyhow::Result<Self>;
}
