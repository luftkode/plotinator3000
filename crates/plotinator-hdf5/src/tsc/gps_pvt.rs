use chrono::{DateTime, Utc};
use hdf5::H5Type;
use ndarray::Array1;
use plotinator_log_if::prelude::*;
use plotinator_ui_util::ExpectedPlotRange;

use crate::tsc::TSC_LEGEND_NAME;

/// Wrapper around all the [`GpsPvtRecord`]s from a TSC.h5 file
pub(crate) struct GpsPvtRecords {
    inner: Array1<GpsPvtRecord>,
}

impl GpsPvtRecords {
    const DATASET_NAME: &str = "GPS_PVT";

    pub fn from_hdf5(h5: &hdf5::File) -> hdf5::Result<Self> {
        let dataset = h5.dataset(Self::DATASET_NAME)?;
        let gps_pvts = dataset.read::<GpsPvtRecord, ndarray::Ix1>()?;
        Ok(Self { inner: gps_pvts })
    }

    // Return a vector of timestamps (unix UTC nanoseconds)
    //
    pub fn timestamps(&self) -> Vec<Option<f64>> {
        self.inner
            .iter()
            .map(|g| g.unix_timestamp_ns().map(|ts| ts as f64))
            .collect()
    }

    #[allow(clippy::too_many_lines, reason = "Long but simple")]
    pub fn build_plots(&self) -> Vec<RawPlot> {
        let time = self.timestamps();

        let mut timestamps = Vec::with_capacity(time.len());
        let mut height = Vec::with_capacity(time.len());
        let mut gspeed = Vec::with_capacity(time.len());
        let mut heading = Vec::with_capacity(time.len());
        let mut lat_deg = Vec::with_capacity(time.len());
        let mut lon_deg = Vec::with_capacity(time.len());

        let mut h_msl = Vec::with_capacity(time.len());
        let mut h_elips = Vec::with_capacity(time.len());
        let mut numsv = Vec::with_capacity(time.len());
        let mut hacc = Vec::with_capacity(time.len());
        let mut vacc = Vec::with_capacity(time.len());
        let mut sacc = Vec::with_capacity(time.len());
        let mut pdop = Vec::with_capacity(time.len());
        let mut vel_n = Vec::with_capacity(time.len());
        let mut vel_e = Vec::with_capacity(time.len());
        let mut vel_d = Vec::with_capacity(time.len());
        let mut mag_dec = Vec::with_capacity(time.len());

        // Valid flags
        let mut valid_date = Vec::with_capacity(time.len());
        let mut valid_time = Vec::with_capacity(time.len());
        let mut valid_fully_resolved = Vec::with_capacity(time.len());
        let mut valid_mag = Vec::with_capacity(time.len());

        // Status flags
        let mut gnss_fix_ok = Vec::with_capacity(time.len());
        let mut diff_soln = Vec::with_capacity(time.len());
        let mut head_veh_valid = Vec::with_capacity(time.len());

        // Additional flags
        let mut confirmed_avail = Vec::with_capacity(time.len());
        let mut confirmed_date = Vec::with_capacity(time.len());
        let mut confirmed_time = Vec::with_capacity(time.len());

        let mut default_hide_valid_date = true;
        let mut default_hide_valid_time = true;
        let mut default_hide_valid_fully_resolved = true;
        let mut default_hide_valid_mag = true;
        let mut default_hide_gnss_fix_ok = true;
        let mut default_hide_diff_soln = true;
        let mut default_hide_head_veh_valid = true;
        let mut default_hide_confirmed_avail = true;
        let mut default_hide_confirmed_date = true;
        let mut default_hide_confirmed_time = true;

        let mut prev_valid_date: Option<f64> = None;
        let mut prev_valid_time: Option<f64> = None;
        let mut prev_valid_fully_resolved: Option<f64> = None;
        let mut prev_valid_mag: Option<f64> = None;
        let mut prev_gnss_fix_ok: Option<f64> = None;
        let mut prev_diff_soln: Option<f64> = None;
        let mut prev_head_veh_valid: Option<f64> = None;
        let mut prev_confirmed_avail: Option<f64> = None;
        let mut prev_confirmed_date: Option<f64> = None;
        let mut prev_confirmed_time: Option<f64> = None;

        for (e, &timestamp_opt) in self.inner.iter().zip(&time) {
            let Some(t) = timestamp_opt else {
                log::warn!("Ignoring NAV-PVT data from invalid timestamp");
                continue;
            };
            timestamps.push(t);
            height.push(e.height_meters());
            gspeed.push(e.ground_speed_kmh());
            heading.push(e.head_mot as f64 * 1e-5); // degrees
            lat_deg.push(e.latitude_degrees());
            lon_deg.push(e.longitude_degrees());

            numsv.push([t, e.num_sv as f64]);
            h_msl.push([t, e.height_msl_meters()]);
            h_elips.push([t, e.height_meters()]);
            hacc.push([t, e.h_acc as f64 * 1e-3]); // mm -> m
            vacc.push([t, e.v_acc as f64 * 1e-3]); // mm -> m
            sacc.push([t, e.s_acc as f64 * 3.6e-3]); // mm/s -> km/h
            pdop.push([t, e.p_dop as f64 * 0.01]);
            vel_n.push([t, e.vel_n as f64 * 3.6e-3]); // mm/s -> km/h
            vel_e.push([t, e.vel_e as f64 * 3.6e-3]); // mm/s -> km/h
            vel_d.push([t, e.vel_d as f64 * 3.6e-3]); // mm/s -> km/h
            mag_dec.push([t, e.mag_dec as f64 * 1e-2]); // degrees * 1e-2 -> degrees

            // Extract valid flags
            let v_valid_date = e.flag_valid(GpsPvtRecord::VALID_DATE);
            let v_valid_time = e.flag_valid(GpsPvtRecord::VALID_TIME);
            let v_valid_fully_resolved = e.flag_valid(GpsPvtRecord::VALID_FULLY_RESOLVED);
            let v_valid_mag = e.flag_valid(GpsPvtRecord::VALID_MAG);
            // Extract status flags
            let v_gnss_fix_ok = e.flag_status(GpsPvtRecord::FLAGS_GNSS_FIX_OK);
            let v_diff_soln = e.flag_status(GpsPvtRecord::FLAGS_DIFF_SOLN);
            let v_head_veh_valid = e.flag_status(GpsPvtRecord::FLAGS_HEAD_VEH_VALID);
            // Extract additional flags
            let v_confirmed_avail = e.flag_additional(GpsPvtRecord::FLAGS2_CONFIRMED_AVAILABLE);
            let v_confirmed_date = e.flag_additional(GpsPvtRecord::FLAGS2_CONFIRMED_DATE);
            let v_confirmed_time = e.flag_additional(GpsPvtRecord::FLAGS2_CONFIRMED_TIME);

            valid_date.push([t, v_valid_date]);
            valid_time.push([t, v_valid_time]);
            valid_fully_resolved.push([t, v_valid_fully_resolved]);
            valid_mag.push([t, v_valid_mag]);
            gnss_fix_ok.push([t, v_gnss_fix_ok]);
            diff_soln.push([t, v_diff_soln]);
            head_veh_valid.push([t, v_head_veh_valid]);
            confirmed_avail.push([t, v_confirmed_avail]);
            confirmed_date.push([t, v_confirmed_date]);
            confirmed_time.push([t, v_confirmed_time]);

            macro_rules! detect_change {
                ($prev:ident, $curr:expr, $flag:ident) => {{
                    if $flag {
                        if let Some(p) = $prev {
                            if (p - $curr).abs() > 0.1 {
                                $flag = false;
                            }
                        }
                        $prev = Some($curr);
                    }
                }};
            }

            detect_change!(prev_valid_date, v_valid_date, default_hide_valid_date);
            detect_change!(prev_valid_time, v_valid_time, default_hide_valid_time);
            detect_change!(
                prev_valid_fully_resolved,
                v_valid_fully_resolved,
                default_hide_valid_fully_resolved
            );
            detect_change!(prev_valid_mag, v_valid_mag, default_hide_valid_mag);
            detect_change!(prev_gnss_fix_ok, v_gnss_fix_ok, default_hide_gnss_fix_ok);
            detect_change!(prev_diff_soln, v_diff_soln, default_hide_diff_soln);
            detect_change!(
                prev_head_veh_valid,
                v_head_veh_valid,
                default_hide_head_veh_valid
            );
            detect_change!(
                prev_confirmed_avail,
                v_confirmed_avail,
                default_hide_confirmed_avail
            );
            detect_change!(
                prev_confirmed_date,
                v_confirmed_date,
                default_hide_confirmed_date
            );
            detect_change!(
                prev_confirmed_time,
                v_confirmed_time,
                default_hide_confirmed_time
            );
        }

        let geo_data: Option<RawPlot> = GeoSpatialDataBuilder::new(TSC_LEGEND_NAME.to_owned())
            .timestamp(&timestamps)
            .lat(&lat_deg)
            .lon(&lon_deg)
            .altitude_from_gnss(height)
            .speed(&gspeed)
            .heading(&heading)
            .build_into_rawplot()
            .expect("invalid builder");

        #[rustfmt::skip]
        let plots = RawPlotBuilder::new(TSC_LEGEND_NAME)
            .add(numsv, DataType::other_unitless("Satellites", ExpectedPlotRange::Hundreds, false))
            .add(h_msl, DataType::AltitudeMSL)
            .add(h_elips, DataType::AltitudeEllipsoidal)
            .add(hacc, DataType::other_distance("Horizontal acc", true))
            .add(vacc, DataType::other_distance("Vertical acc", true))
            .add(sacc, DataType::other_velocity("Speed acc", true))
            .add(pdop, DataType::other_unitless("PDOP", ExpectedPlotRange::Hundreds, true))
            .add(vel_n, DataType::other_velocity("Velocity North", true))
            .add(vel_e, DataType::other_velocity("Velocity East", true))
            .add(vel_d, DataType::other_velocity("Velocity Down", true))
            .add(mag_dec, DataType::other_degrees("Magnetic Decl.", true))
            .add(valid_date, DataType::bool("Valid Date", default_hide_valid_date))
            .add(valid_time, DataType::bool("Valid Time", default_hide_valid_time))
            .add(valid_fully_resolved, DataType::bool("Valid fully resolved", default_hide_valid_fully_resolved))
            .add(valid_mag, DataType::bool("Valid Magnetic Declination", default_hide_valid_mag))
            .add(gnss_fix_ok, DataType::bool("GNSS Fix OK", default_hide_gnss_fix_ok))
            .add(diff_soln, DataType::bool("Differential Solution", default_hide_diff_soln))
            .add(head_veh_valid, DataType::bool("Head Vehicle Valid", default_hide_head_veh_valid))
            .add(confirmed_avail, DataType::bool("Confirmed Available", default_hide_confirmed_avail))
            .add(confirmed_date, DataType::bool("Confirmed Date", default_hide_confirmed_date))
            .add(confirmed_time, DataType::bool("Confirmed Time", default_hide_confirmed_time));
        let mut plots = plots.build();

        if let Some(geo_data) = geo_data {
            plots.push(geo_data);
        }
        plots
    }
}
/// UBX-NAV-PVT message payload structure
#[derive(H5Type, Debug)]
#[repr(C)]
struct GpsPvtRecord {
    /// GPS time of week of navigation epoch (milliseconds)
    #[hdf5(rename = "iTOW")]
    i_tow: u32,

    /// Year (UTC)
    year: u16,
    /// Month (UTC), range 1-12
    month: u8,
    /// Day of month (UTC), range 1-31
    day: u8,
    /// Hour of day (UTC), range 0-23
    hour: u8,
    /// Minute of hour (UTC), range 0-59
    min: u8,
    /// Seconds of minute (UTC), range 0-60 (60 for leap seconds)
    sec: u8,

    /// Validity flags
    /// Bit 0 (`VALID_DATE)`: valid UTC date
    /// Bit 1 (`VALID_TIME)`: valid UTC time of day
    /// Bit 2 (`VALID_FULLY_RESOLVED)`: UTC time fully resolved (no seconds uncertainty)
    /// Bit 3 (`VALID_MAG)`: valid magnetic declination
    #[hdf5(rename = "validFlag")]
    valid_flag: u8,

    /// Time accuracy estimate (nanoseconds)
    #[hdf5(rename = "tAcc")]
    t_acc: u32,

    /// Fraction of second, range -1e9..1e9 (UTC) (nanoseconds)
    nano: i32,

    /// GNSS fix type, range 0-5
    /// 0: No fix
    /// 1: Dead reckoning only
    /// 2: 2D-fix (signal from only 3 SVs, constant altitude assumed)
    /// 3: 3D-fix
    /// 4: GNSS + dead reckoning combined
    /// 5: Time only fix (high precision devices)
    #[hdf5(rename = "fixType")]
    fix_type: u8,

    /// Fix status flags
    /// Bit 0 (`FLAGS_GNSS_FIX_OK)`: valid fix (within DOP & accuracy masks)
    /// Bit 1 (`FLAGS_DIFF_SOLN)`: DGPS used
    /// Bits 2-4 (`FLAGS_PSM_MASK)`: Power save mode state
    /// Bit 5 (`FLAGS_HEAD_VEH_VALID)`: heading of vehicle is valid
    /// Bits 6-7 (`FLAGS_CARRIER_PHASE_MASK)`: Carrier phase solution status
    flags: u8,

    /// Additional flags
    /// Bit 5 (`FLAGS2_CONFIRMED_AVAILABLE)`: UTC Date/Time validity confirmation available
    /// Bit 6 (`FLAGS2_CONFIRMED_DATE)`: UTC Date validity confirmed
    /// Bit 7 (`FLAGS2_CONFIRMED_TIME)`: UTC Time of Day validity confirmed
    flags2: u8,

    /// Number of satellites used in navigation solution
    #[hdf5(rename = "numSV")]
    num_sv: u8,

    /// Longitude (degrees * 1e-7)
    lon: i32,
    /// Latitude (degrees * 1e-7)
    lat: i32,
    /// Height above ellipsoid (millimeters)
    height: i32,
    /// Height above mean sea level (millimeters)
    #[hdf5(rename = "hMSL")]
    h_msl: i32,

    /// Horizontal accuracy estimate (millimeters)
    #[hdf5(rename = "hAcc")]
    h_acc: u32,
    /// Vertical accuracy estimate (millimeters)
    #[hdf5(rename = "vAcc")]
    v_acc: u32,

    /// NED north velocity (millimeters/second)
    #[hdf5(rename = "velN")]
    vel_n: i32,
    /// NED east velocity (millimeters/second)
    #[hdf5(rename = "velE")]
    vel_e: i32,
    /// NED down velocity (millimeters/second)
    #[hdf5(rename = "velD")]
    vel_d: i32,

    /// Ground speed (2-D) (millimeters/second)
    #[hdf5(rename = "gSpeed")]
    g_speed: i32,

    /// Heading of motion (2-D) (degrees * 1e-5)
    #[hdf5(rename = "headMot")]
    head_mot: i32,

    /// Speed accuracy estimate (millimeters/second)
    #[hdf5(rename = "sAcc")]
    s_acc: u32,

    /// Heading accuracy estimate (both motion & vehicle) (degrees * 1e-5)
    #[hdf5(rename = "headAcc")]
    head_acc: u32,

    /// Position DOP (dilution of precision) (1 / 0.01)
    #[hdf5(rename = "pDOP")]
    p_dop: u16,

    reserved01: u8,
    reserved02: u8,
    reserved03: u8,
    reserved04: u8,
    reserved05: u8,
    reserved06: u8,

    /// Heading of vehicle (2-D) (degrees * 1e-5)
    /// Vehicle orientation, different from headMot (direction of motion)
    #[hdf5(rename = "headVeh")]
    head_veh: i32,

    /// Magnetic declination (degrees * 1e-2)
    #[hdf5(rename = "magDec")]
    mag_dec: i16,
    /// Magnetic declination accuracy (degrees * 1e-2)
    #[hdf5(rename = "magAcc")]
    mag_acc: u16,
}

impl GpsPvtRecord {
    // Valid flag constants
    const VALID_DATE: u8 = 1;
    const VALID_TIME: u8 = 2;
    const VALID_FULLY_RESOLVED: u8 = 4;
    const VALID_MAG: u8 = 8;

    // Status flag constants
    const FLAGS_GNSS_FIX_OK: u8 = 1;
    const FLAGS_DIFF_SOLN: u8 = 2;
    const FLAGS_HEAD_VEH_VALID: u8 = 32;

    // Additional flag constants
    const FLAGS2_CONFIRMED_AVAILABLE: u8 = 32;
    const FLAGS2_CONFIRMED_DATE: u8 = 64;
    const FLAGS2_CONFIRMED_TIME: u8 = 128;

    fn unix_timestamp_utc(&self) -> Option<DateTime<Utc>> {
        let Some(naive_date) =
            chrono::NaiveDate::from_ymd_opt(self.year as i32, self.month as u32, self.day as u32)
        else {
            log::error!("Invalid date: {self:?}");
            return None;
        };

        let Some(naive_datetime) =
            naive_date.and_hms_opt(self.hour as u32, self.min as u32, self.sec as u32)
        else {
            log::error!("Invalid time: {self:?}");
            return None;
        };

        let mut utc_datetime = naive_datetime.and_utc();

        // Apply nanosecond correction (could be timing offset/correction)
        if self.nano != 0 {
            utc_datetime += chrono::Duration::nanoseconds(self.nano as i64);
        }

        Some(utc_datetime)
    }

    /// Convert PVT record to precise Unix timestamp in nanoseconds since Unix epoch
    fn unix_timestamp_ns(&self) -> Option<i64> {
        self.unix_timestamp_utc()
            .and_then(|ts| ts.timestamp_nanos_opt())
    }

    /// Get latitude in degrees
    pub fn latitude_degrees(&self) -> f64 {
        self.lat as f64 * 1e-7
    }

    /// Get longitude in degrees
    pub fn longitude_degrees(&self) -> f64 {
        self.lon as f64 * 1e-7
    }

    /// Get height above ellipsoid in meters
    pub fn height_meters(&self) -> f64 {
        self.height as f64 * 1e-3
    }

    /// Get height above mean sea level in meters
    pub fn height_msl_meters(&self) -> f64 {
        self.h_msl as f64 * 1e-3
    }

    /// Get ground speed in km/h
    pub fn ground_speed_kmh(&self) -> f64 {
        self.g_speed as f64 * 3.6e-3
    }

    /// Extract flag from `valid_flag` field as float (0.0 or 1.0)
    fn flag_valid(&self, mask: u8) -> f64 {
        if self.valid_flag & mask != 0 {
            1.0
        } else {
            0.0
        }
    }

    /// Extract flag from flags field as float (0.0 or 1.0)
    fn flag_status(&self, mask: u8) -> f64 {
        if self.flags & mask != 0 { 1.0 } else { 0.0 }
    }

    /// Extract flag from flags2 field as float (0.0 or 1.0)
    fn flag_additional(&self, mask: u8) -> f64 {
        if self.flags2 & mask != 0 { 1.0 } else { 0.0 }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use chrono::NaiveDate;
    use plotinator_test_util::test_file_defs::tsc::*;
    use testresult::TestResult;

    #[test]
    fn read_gps_pvt() -> TestResult {
        let h5file = hdf5::File::open(tsc())?;
        let gps_pvt = GpsPvtRecords::from_hdf5(&h5file)?;
        let gps_pvt = gps_pvt.inner;
        for g in &gps_pvt {
            eprintln!("g: {g:?}");
            eprintln!("{:?}", g.unix_timestamp_utc());
            eprintln!(
                "Lat: {:.7}°, Lon: {:.7}°",
                g.latitude_degrees(),
                g.longitude_degrees()
            );
            eprintln!(
                "Height: {:.3}m, Speed: {:.3}m/s",
                g.height_meters(),
                g.ground_speed_kmh()
            );
        }

        insta::assert_debug_snapshot!(gps_pvt);
        Ok(())
    }

    #[test]
    fn read_gps_pvt_from_stub() -> TestResult {
        let h5file = hdf5::File::open(tsc_stub())?;
        let gps_pvt = GpsPvtRecords::from_hdf5(&h5file)?;
        let gps_pvt = gps_pvt.inner;
        for g in &gps_pvt {
            eprintln!("g: {g:?}");
            eprintln!("{:?}", g.unix_timestamp_utc());
            eprintln!(
                "Lat: {:.7}°, Lon: {:.7}°",
                g.latitude_degrees(),
                g.longitude_degrees()
            );
            eprintln!(
                "Height: {:.3}m, Speed: {:.3}m/s",
                g.height_meters(),
                g.ground_speed_kmh()
            );
        }

        insta::assert_debug_snapshot!(gps_pvt);
        Ok(())
    }

    #[test]
    fn test_timestamp_conversion() {
        let record = GpsPvtRecord {
            i_tow: 345600000, // 4 days into week in milliseconds
            year: 2022,
            month: 3,
            day: 10,
            hour: 15,
            min: 30,
            sec: 45,
            valid_flag: 0x07, // Valid date/time
            t_acc: 100,
            nano: 123456789, // 123.456789 milliseconds additional nanoseconds
            fix_type: 3,
            flags: 0,
            flags2: 0,
            num_sv: 12,
            lon: 550_000_000, // 55.0 degrees * 1e7
            lat: 125_000_000, // 12.5 degrees * 1e7
            height: 100_000,  // 100m * 1e3
            h_msl: 50_000,    // 50m * 1e3
            h_acc: 2000,
            v_acc: 3000,
            vel_n: 1000,     // 1 m/s * 1e3
            vel_e: 500,      // 0.5 m/s * 1e3
            vel_d: -200,     // -0.2 m/s * 1e3
            g_speed: 2000,   // 2 m/s * 1e3
            head_mot: 90000, // 90 degrees * 1e3
            s_acc: 100,
            head_acc: 5000,
            p_dop: 180, // 1.8 * 100
            reserved01: 0,
            reserved02: 0,
            reserved03: 0,
            reserved04: 0,
            reserved05: 0,
            reserved06: 0,
            head_veh: 90000,
            mag_dec: 0,
            mag_acc: 0,
        };

        // Verify the expected datetime
        assert_eq!(
            record.unix_timestamp_utc().unwrap(),
            NaiveDate::from_ymd_opt(2022, 3, 10)
                .unwrap()
                .and_hms_nano_opt(15, 30, 45, 123456789)
                .unwrap()
                .and_utc()
        );
    }
}
