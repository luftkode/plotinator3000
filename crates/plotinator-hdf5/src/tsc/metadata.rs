use crate::util;

#[derive(Debug)]
pub(crate) struct RootMetadata {
    linear_time: f64,
    kickin_usec: f64,
    hm_ontime_usec: f64,
    gates_pr_decade_on: f64,
    metadata: Vec<(String, String)>,
}

impl RootMetadata {
    pub fn parse_from_tsc(h5: &hdf5::File) -> anyhow::Result<Self> {
        let mut metadata = vec![];
        for attr_name in h5.attr_names()? {
            let attr = h5.attr(&attr_name)?;
            let attr_str = util::read_any_attribute_to_string(&attr)?;
            metadata.push((attr_name, attr_str));
        }

        let linear_time = h5.attr("t_lin")?.read_scalar::<f64>()?;
        let kickin_usec = h5.attr("kickin_usec")?.read_scalar::<f64>()?;
        let hm_ontime_usec = h5.attr("hm_ontime_usec")?.read_scalar::<f64>()?;
        let gates_pr_decade_on = h5.attr("gates_pr_decade_on")?.read_scalar()?;

        Ok(Self {
            linear_time,
            kickin_usec,
            hm_ontime_usec,
            gates_pr_decade_on,
            metadata,
        })
    }
    /// 0-based
    pub(crate) fn last_gate_on_index(&self) -> usize {
        self.last_gate_on_count() - 1
    }
    /// 1-based (count)
    pub(crate) fn last_gate_on_count(&self) -> usize {
        let k_ramp_up = gates_pr_decade_to_k(self.gates_pr_decade_on);
        let k_ontime = gates_pr_decade_to_k(self.gates_pr_decade_on);
        let gate_kick_in = sinh_gates(self.kickin_usec, self.linear_time, k_ramp_up);

        let gate_from_kick_in_to_ramp_down: Vec<f64> = sinh_gates(
            self.hm_ontime_usec - self.kickin_usec,
            self.linear_time,
            k_ontime,
        )
        .into_iter()
        .map(|x| x + self.kickin_usec)
        .collect();

        gate_kick_in.len() + gate_from_kick_in_to_ramp_down.len()
    }

    pub(crate) fn metadata_strings(&self) -> Vec<(String, String)> {
        self.metadata.clone()
    }
}

fn gates_pr_decade_to_k(gpd: f64) -> f64 {
    1.0 / (gpd * std::f64::consts::E.log10())
}

fn sinh_gates(t: f64, t_lin: f64, k: f64) -> Vec<f64> {
    // Calculate n_max = ceil((1/k) * arcsinh(t / t_lin))
    let n_max = ((1.0 / k) * (t / t_lin).asinh()).ceil() as usize;

    // Build tgates = [t_lin * sinh(k * n) for n in 1..=n_max]
    let mut tgates: Vec<f64> = (1..=n_max)
        .map(|n| t_lin * (k * (n as f64)).sinh())
        .collect();

    // Replace last element with t
    if let Some(last) = tgates.last_mut() {
        *last = t;
    }

    tgates
}

#[cfg(test)]
mod tests {
    use plotinator_test_util::test_file_defs::tsc::tsc;
    use testresult::TestResult;

    use super::*;

    #[test]
    fn test_read_tsc_attributes() -> TestResult {
        let h5file = hdf5::File::open(tsc())?;

        let root_metadata = RootMetadata::parse_from_tsc(&h5file)?;

        let last_on = root_metadata.last_gate_on_count();
        assert_eq!(last_on, 15);

        Ok(())
    }
}
