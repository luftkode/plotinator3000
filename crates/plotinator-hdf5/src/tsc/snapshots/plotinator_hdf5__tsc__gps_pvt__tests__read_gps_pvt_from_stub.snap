---
source: crates/plotinator-hdf5/src/tsc/gps_pvt.rs
expression: gps_pvt
---
[GpsPvtRecord {
    i_tow: 0,
    year: 0,
    month: 0,
    day: 0,
    hour: 0,
    min: 0,
    sec: 0,
    valid_flag: 0,
    t_acc: 0,
    nano: 0,
    fix_type: 0,
    flags: 0,
    flags2: 0,
    num_sv: 0,
    lon: 0,
    lat: 0,
    height: 0,
    h_msl: 0,
    h_acc: 0,
    v_acc: 0,
    vel_n: 0,
    vel_e: 0,
    vel_d: 0,
    g_speed: 0,
    head_mot: 0,
    s_acc: 0,
    head_acc: 0,
    p_dop: 0,
    reserved01: 0,
    reserved02: 0,
    reserved03: 0,
    reserved04: 0,
    reserved05: 0,
    reserved06: 0,
    head_veh: 0,
    mag_dec: 0,
    mag_acc: 0,
}], shape=[1], strides=[1], layout=CFcf (0xf), const ndim=1
